UniStuttgart-VISUS/megamol
# ClipPlane GUI Widget

Added the imguizo library to provide an interactive manipulator widget for the clipping plane module.

The ImGui part is realized as parameter group widget in the GUI frontend. 
Callers have to send the current camera first, before they request the clip plane via the callback of the ClipPlaneCall.

TODO
- [ ] Rename ClipPlane module to PlaneModule?
- [ ] Rename ClipPlaneCall call to PlaneCall?

ISSUES
- [ ] Widget gets clipped at internal near plane that is way too near....

FEATURES
- [ ] button to invert the normal of the plane to switch between clipping "in front" and "behind"
- [ ] separate between "enabled" (does some clipping) and "manipulatable" (widgets are shown and can be moved); maybe add option under draw to enable/disable drawing the dragable widgets


Already contains changes of PR #1195 and #1194 ...

## Summary of Changes


## References and Context


## Test Instructions

- Already implemented usage: See changes in SphereRenderer ...
- Add the following code to your renderer in Render():

```C++
    // Requires call slot 'get_clip_plane_slot_':
    // this->get_clip_plane_slot_.SetCompatibleCall<view::CallClipPlaneDescription>();
    // this->MakeSlotAvailable(&this->get_clip_plane_slot_); 
    megamol::core::view::Camera cam = call.GetCamera();
    view::CallClipPlane* ccp = this->get_clip_plane_slot_.CallAs<view::CallClipPlane>();
    if (ccp != nullptr) {
        ccp->SetCamera(cam);
    }
    if ((ccp != nullptr) && (*ccp)()) {
        auto plane = ccp->GetPlane();
    }
```
